pkg/mdformatter/linktransformer/link.go

```diff
@@ -175,14 +180,53 @@ func NewValidator(logger log.Logger, linksValidateConfig []byte, anchorDir strin
 	v.c.OnError(func(response *colly.Response, err error) {
 		v.rMu.Lock()
 		defer v.rMu.Unlock()
-		v.remoteLinks[response.Ctx.Get(originalURLKey)] = errors.Wrapf(err, "%q not accessible; status code %v", response.Request.URL.String(), response.StatusCode)
+		retriesStr := response.Ctx.Get(numberOfRetriesKey)
+		retries, _ := strconv.Atoi(retriesStr)
+		switch response.StatusCode {
+		case http.StatusTooManyRequests:
+			if retries > 0 {
+				break
+			}
+			var retryAfterSeconds int
+			// Retry calls same methods as Visit and makes request with same options.
+			// So retryKey is incremented here if onError is called again after Retry. By default retries once.
+			response.Ctx.Put(numberOfRetriesKey, strconv.Itoa(retries+1))
+			retryAfterSeconds, convErr := strconv.Atoi(response.Headers.Get("Retry-After"))
+			if convErr != nil {
+				retryAfterSeconds = 1
+			}
+			select {
+			case <-time.After(time.Duration(retryAfterSeconds) * time.Second):
+			case <-v.c.Context.Done():
+				return
+			}
+
+			if retryErr := response.Request.Retry(); retryErr != nil {
+				v.remoteLinks[response.Ctx.Get(originalURLKey)] = errors.Wrapf(err, "remote link retry %v", response.Ctx.Get(originalURLKey))
+				break
+			}
+			v.remoteLinks[response.Ctx.Get(originalURLKey)] = errors.Wrapf(err, "%q rate limited even after retry; status code %v", response.Request.URL.String(), response.StatusCode)
+		case http.StatusMovedPermanently, http.StatusTemporaryRedirect, http.StatusServiceUnavailable:
+			if retries > 0 {
+				break
+			}
+			response.Ctx.Put(numberOfRetriesKey, strconv.Itoa(retries+1))
+
+			if retryErr := response.Request.Retry(); retryErr != nil {
+				v.remoteLinks[response.Ctx.Get(originalURLKey)] = errors.Wrapf(err, "remote link retry %v", response.Ctx.Get(originalURLKey))
+				break
+			}
+			v.remoteLinks[response.Ctx.Get(originalURLKey)] = errors.Wrapf(err, "%q not accessible even after retry; status code %v", response.Request.URL.String(), response.StatusCode)
+		default:
+			v.remoteLinks[response.Ctx.Get(originalURLKey)] = errors.Wrapf(err, "%q not accessible; status code %v", response.Request.URL.String(), response.StatusCode)
+		}
 	})
 	return v, nil
 }
 
 // MustNewValidator returns mdformatter.LinkTransformer that crawls all links.
 func MustNewValidator(logger log.Logger, linksValidateConfig []byte, anchorDir string) mdformatter.LinkTransformer {
-	v, err := NewValidator(logger, linksValidateConfig, anchorDir)
+	v, err := NewValidator(context.TODO(), logger, linksValidateConfig, anchorDir)
```

Comment:
does not sound useful ;p

Reply:
We can currently interrupt retry. For instance, testing a file with envoyproxy link and setting retryAfterSeconds to 10, gives the following results. First is when I let it run and retry, second is when I interrupt,

```
time mdox fmt -l test.md
error:          test.md: "https://envoyproxy.io" not accessible even after retry; status code 301: Moved Permanently
mdox fmt -l test.md  0.14s user 0.06s system 1% cpu 14.995 total

ime mdox fmt -l test.md
^Cinfo: caught signal. Exiting.: interrupt
mdox fmt -l test.md  0.12s user 0.05s system 2% cpu 6.762 total
```
So we can cancel retry without having to wait. But I'm not sure what would be the correct way to hook context?

Reply:
Yes, we pass it [here](https://github.com/bwplotka/mdox/blob/main/pkg/mdformatter/linktransformer/link.go#L158)! That's why we can cancel without waiting.